### DISCUSSION TITLE
Tokens on OpenSea are not showing after contract deploy

### DISCUSSION THREAD

**Discussion by ecerroni** [hashlips-lab/nft-erc721-collection#210] [General]
There seems to be a recurring issue in getting the tokens being displayed on OpenSea after successfully deploying the contract.

Unfortunately, I've been of those that experienced this before hand. I tried to investigate the issue in many ways as I wanted to check if it was something related to the contract or because of OpenSea.

My conclusions are that **the contract in its current implementation is totally fine**. The issue seems to be with OpenSea and how it handles new contracts deployed on the block chain. 

Generally the issues the user may facing are:
- error messages while trying importing the contract: _"We couldn't find this contract. Please ensure that this is a valid ERC721 or ERC1155 contract deployed on Ethereum/Polygon and that you have already minted items on the contract."_
- Contract has been imported, but none or less tokens are listed

This is mainly on OpenSea side. I found an interesting article that sums it up quite well:

-----------------------
Based on observed behavior this is how OpenSea ERC721 validation works:

1. OpenSea automatically monitors all new contracts that are deployed. The moment OpenSea detects a new contract has been deployed it calls contract.supportsInterface(type(IERC721).interfaceId) on that contract. If the function call returns true then OpenSea adds it to its list or database of potential ERC721 contracts. If the function call returns false then OpenSea will not detect it as an ERC721 contract and any attempts at validation fails.

2. OpenSea checks during deployment and after deployment (for any contract that passes step 1 above) if any ERC721 Transfer events have been emitted from a contract. If not then any validation fails.

If a contract passes step 1 and 2 then it will pass ERC721 validation on

-----------------------

ref: https://dev.to/mudgen/how-to-solve-openseas-we-couldnt-find-this-contract-during-erc721-validation-4lp2

There are many issues that people opened on this repo that claim having these kind of problem. Both Ethereum and Polygon may be affected with Polygon being significantly more vulnerable to this. Personally I deployed on Polygon when I had this issue.

The fact that successfully deploying the contract to testnets first does not prevent this kind of issues on the mainet.

As @liarco himself pointed out there is no rule here, but a single mint operation with thousands of tokens can have negative impact on the expected outcome.

My 0.02 is that:
- OpenSea db update cannot catch up with thousand of events firing at the same time for a single contract, it throws an error and the minting events are gone forever from their point of view
- OpenSea does not like thousands of concurrent mints at the same time in the first place and it just ignore them. Maybe it sees is as spam.
- OpenSea/Blockchain integration is far from perfect and things get lost with this kind of excessive load, especially on Polygon

The above assumptions are based on my personal circumstantial evidence:
1. I deployed the contract
2. I minted 4947 items
3. I cannot import the contract (I tried twice a day)
4. In the meantime contract is imported on Rarible (after 6 days of failed attempts), Refinable (although with the help of their devs) and NFTrade (quite quickly) with all 4947 items showing properly.
5. after 2 weeks I mint an additional item
6. I was able to finally import the contract on OpenSea with one item showing (the last mint only)
7. nothing changed for 3 days
8. I did transfer one of the initial minted items from me to myself again. This item showed up on OpenSea under the collection after just a few minutes.

So the main takeaway is to **not mint all at once**. I am also keen to believe that with the standard process of minting through the dapp will not result in any issue.

Any contribution to this discussion is very welcome.

**Comment by liarco** [comment]
Hi @ecerroni, thank you so much for taking time to share this detailed explanation. This is an extremely valuable piece of information.

I'm gonna pin this discussion for a while so people will be able to see it immediately.

**Comment by FrancoGedor** [comment]
This is really interesting. I also had this issue with Polygon when minting 10 or more NFTs at once. I opened a ticket with OpenSea to see if they can resolve this somehow.

I wonder if there's a "magic number" of NFTs that you can mint without running into this problem. I will do some tests today and post back with the results.

**Reply by ShehanJay01** [reply]
Did you find a "magic number"?

**Comment by dcryptodev** [comment]
Thank you @ecerroni, perfect explanation, since ive been using this project i havent had a specific bug related issue more issues i've had were all on Opensea, I opened a ticket for the same exact problem. It definitely is an OpenSea issue that needs resolution from their side. I am sure Opensea is well aware of this issue.

**Comment by Angrywolvesdev** [comment]
Has anyone been able to solve this problem? The only thing I have been able to do is mint 5 at a time, which doesn't make sense if you want to make a collection of 5,000 or more pieces. I tried 10 in 10 but the tokens do not appear in OpenSea, it only allows 5 since it is a small number and easy to load I guess

**Reply by ecerroni** [reply]
@Angrywolvesdev I've been also thinking on how to solve that for a while. I tried to find a solution that would allow me to mint all at once, but to fire the events one by one every second.

However, afaik Solidity does not have the typical javascript features like async/await, setInterval, setTimeout, etc. I am still in the initial stage of learning Solidity, but from what I found that is not possible.

I thought also to use hardhat for automating this task, but that would mean waiting to click on Metamask 5.000 times :D

It's true that we could automate it with Selenium or other tools like that, but it's not so easy to implement.

Last thing that came to my mind, but I did not find the time to try, is to attempt bulk transfers. I am wondering if bulk transfers would be ignored by OpenSea like bulk mints 🤔 

Indeed mint puts the token on-chain. However, a transfer executes on a token being already on chain. Maybe OpenSea will like that more.

What do you think?

**Comment by Angrywolvesdev** [comment]
I was testing and today I managed to mint the entire collection with an automatic program. Since the contract in Polyscan in "mint to address" I put only 8 in quantity and the address that I wanted to upload the collection, it takes a while to load but it was successful. I recommend not to place more than 8 because they will no longer appear in OpenSea. I also recommend not doing it too quickly, as they may not show up in as many transfers. Charging took about three hours. If this helped anyone, you can take a look at my collection: 

https://opensea.io/collection/angrywolvesclub

**Reply by Angrywolvesdev** [reply]
By the way, the program is called Mouse Recorder pro 2, it is very easy to use and without programming, although there are many other macros, that was the one I used.

**Comment by Angrywolvesdev** [comment]
Yes, I do.  8 is the maximum amount per transaction to appear in Opensea, a
higher number will not appear in the collection

**Reply by dcryptodev** [reply]
@Angrywolvesdev great feedback and i will be taking this onboard

 I have two questions based on what you did:

1. So are you saying that when someone mints an NFT the max they should mint for it to always show up on opensea is 8 for Polygon? also is there a specific **wait time** before minting next 8 NFTs from the supply? did you wait at all did you do it instantly after the first 8. 

2. Lets say you wanted a group of people to mint maximum 2 NFT's from a total of 4000 NFT just for the whitelisting sale how would you go about doing it to make sure it always shows up on Opensea and the NFTs don't go missing. 

Thanks

**Reply by zoultrex** [reply]
@Angrywolvesdev is this behavior (on Opensea) only for mainnet or can you see this also on testnet?

**Reply by ecerroni** [reply]
Testnet/OpenSea is much more forgiving. Depending on the traffic at the time I was once able mint 1000 tokens at a time.

No way this could ever happen on Mainet.

**Comment by Angrywolvesdev** [comment]
Exactly, only 8 at a time.  At first I did it by clicking by myself and at
the same time looking at the transfers of the contract in Polygonscan, the
items appear in Opensea 1 minute after being minted.  Later, to speed up
the process, use a macro with a normal speed, even with about 5 seconds of
waiting between transfers, all so as not to saturate and that the Nfts
appear in Opensea.

I hope it helps you, I would love to see what you created⚡️⚡️ greetings 👋🏼

**Reply by dcryptodev** [reply]
I really appreciate your help this is awesome i will be doing that going forward. The only thing i'm thinking about is how do you force people to wait for at least 5 seconds when a whitelisting sale is happening from the dApp so to stop everyone minting too fast, then it doesn't show up on openSea. Maybe it be good to set a 5-10 seconds the moment the users join the website. 

Also wanted to ask were you able to figure out why your NFTs were showing up in **new traits** issue and the properties ? i see your NFT's are not showing the percentages yet. 

 I will post the Nft here for sure when we push to opensea im still currently on testnet 😆

**Comment by Angrywolvesdev** [comment]
In my case, since I decided to mint the entire collection, I did what you
mentioned.  but if many people are interested in your project and want to
do it from the dapp, I don't think there will be a problem, you should
limit it to 8 nfts per person at most, but when they are minted in
different directions the time problem does not happen.  The percentage of
traits is just a matter of waiting.  I hope you have understood me, by the
way I would like to see your collection.

**Comment by Angrywolvesdev** [comment]
in tesnet this does not happen, it is a test and there you can mint the
entire collection

**Comment by ecerroni** [comment]
After 2 months the saga is finally over and I am very happy to post here my findings.

TLDR; OpenSea's blockchain spam filter will probably flag the mass-minting and it will not import the NFTs from the collection.

I contacted OpenSea support on 18.3.22 and a few days ago they fixed the issue of my collection that was not showing any of the initial NFTs I minted all at once.

The issue was finally solved by a support operator that had backfilled the minting transaction I provided in my initial request and updated the stats.

I insisted in getting an explanation of the reason why that happened and here it is:

-------------------
So it appears that our blockchain spam filter (the version implemented at the time) was a little too aggressive in this case. Transactions that include the minting of several thousand NFTs at once were sometimes being flagged as spam and then ignored by our blockchain scanner. Our blockchain scanner is what ingests blockchain data that's later organized and displayed on OpenSea. Sometimes a few NFTs from the transaction would slip through before the spam filter would kick in to block the rest. It was fairly inconsistent for a little while which prompted us to make some major improvements that have since been implemented.

I'm really sorry that your collection got caught up in the mix here!

------------------

This is what I was suspecting after a few weeks investigating the issue and now it's been confirmed by their support service.

My 0,02$: never mint NFTs of your collection all at once, even if their filter has been updated!

**Comment by Hofstud** [comment]
Thanks so much for this thread as I am having the same issue. I contacted OpenSea support who confirmed its a bug on their side. M ticket is now with the dev team to sort out. (8June22)

**Reply by liarco** [reply]
Thank you very much for letting us know. I would really appreciate it if you could follow up on this in case you have further news. 😉

**Reply by Hofstud** [reply]
I am still chasing up this error with OpenSea and this is their latest reply...this is from the 2nd support person as I raised a 2nd ticket to try to elevate the request as its been a month now

"
Hey there,
 
Thanks for reaching out and alerting us about this. After some testing, this does appear to be a bug on OpenSea. I’ve escalated this to our Developer team and they are actively investigating this for you.
 
As a quick note, depending on the complexity of the issue, it takes a bit of time to find a resolution. I’m sorry to say, we can’t provide a timeline on when this will be fixed.
 
With that said, I know offering a solution is really important here. I’m going to keep this case pending, and follow up with you once I get confirmation from our Developers that this has been fixed. I’ll keep my fingers crossed that I can give you an update soon.
 
Again, thank you for your patience and I’ll be in touch as soon as I can.
"

**Comment by barakacap** [comment]


**Reply by liarco** [reply]
@barakacap thank you for reporting your experience with this, but IMHO saying that the `ERC721A` is "the problem" here is a bit of a strong statement, especially considering the huge impact that using plain `ERC721` has on the economy of a collection.

It's known that swapping the base contract avoids the issue (as well as the benefits of the Azuki's implementations and this is why we don't see this as a solution), but there are also direct statements from OpenSea that the issue is with how they are handling the thing on their side. The Azuki's contract doesn't seem to have any real bug related to this.

If anyone is willing to sacrifice gas efficiency for this, that's fine, but it's important to know that there's no evidence that the `ERC721A` contract has any issue.

Thank you!

**Comment by mrmozaffari** [comment]
I have the exact same issue with importing contracts into OpenSea. In my case, I was able to import the contract for the first time and then deleted the collection from OpenSea. Once I tried to re-import again it failed with this error:

**"We couldn't find this contract. Please ensure that this is a valid ERC721 or ERC1155 contract deployed on Polygon and that you have already minted items on the contract."**

I'm able to import the contract and see the items in different Market-Places so it seems it's an issue on OpenSea. I opened a ticket and got this response and they claimed the issue has been fixed. It has not! I also checked our MetaData and all is there. Typically the MetaData needs to have the following object:

Example:

{
  "description": "Friendly OpenSea Creature that enjoys long swims in the ocean.", 
  "external_url": "https://openseacreatures.io/3", 
  "image": "https://storage.googleapis.com/opensea-prod.appspot.com/puffs/3.png", 
  "name": "Dave Starbelly",
  "attributes": [ ... ], 
}

OpenSea response to the ticket:

> "**I had our team manually refresh the contract creation and 10 minting even transactions. This issue should now be resolved. You can try re-importing the contract again.**"

The issue is still there.

Thank you!

**Comment by Hofstud** [comment]
Good Morning Everyone,

I have finally resolved the issue with my collection not showing on Opensea and for my case it had to do with pinning the collection to Pinata as having the collection pinned locally to my IPFS node just didn't seem to work. 

I can now see my collection on opensea test net so now have the confidence to launch the MainNet Contract.

I am so grateful for all the support and learning from Daniel and the Hashlips community

Below is a screenshot of the test with my Friesian babies collection launching on 22nd July 2022

https://hofs.io/

<img width="1670" alt="Screen Shot 2022-07-10 at 8 13 44 am" src="https://user-images.githubusercontent.com/96272109/178126773-bec3c5fe-ae64-43a8-b528-bab5f2ea67aa.png">
